File: cpp/perceptron/Perceptron.cpp
#include "Perceptron.h"
#include <algorithm>
#include <numeric>
#include <cmath>

Perceptron::Perceptron(int n_inputs, double learning_rate, int max_epochs, int random_state)
    : n_inputs_(n_inputs), 
      learning_rate_(learning_rate), 
      max_epochs_(max_epochs),
      random_state_(random_state),
      bias_(0.0) 
{
    // Initialize random number generator with seed
    rng_.seed(random_state_);
    
    // Initialize weights with small random values
    std::normal_distribution<double> dist(0.0, 0.01);
    weights_.resize(n_inputs_);
    for (auto& w : weights_) {
        w = dist(rng_);
    }
}

Perceptron& Perceptron::fit(const std::vector<std::vector<double>>& X, const std::vector<int>& y) {
    // Reset errors list
    errors_.clear();
    
    // Training loop
    for (int epoch = 0; epoch < max_epochs_; ++epoch) {
        // Make predictions
        std::vector<int> y_pred = predict(X);
        
        // Update weights and get misclassification count
        int misclassified = update_weights(X, y, y_pred);
        errors_.push_back(misclassified);
        
        // Stop training if all samples are correctly classified
        if (misclassified == 0) {
            break;
        }
    }
    
    return *this;
}

std::vector<int> Perceptron::predict(const std::vector<std::vector<double>>& X) {
    std::vector<int> predictions;
    predictions.reserve(X.size());
    
    for (const auto& x : X) {
        // Calculate net input: w·x + b
        double net_input = bias_;
        for (size_t i = 0; i < x.size(); ++i) {
            net_input += weights_[i] * x[i];
        }
        
        // Apply step activation function
        predictions.push_back(net_input >= 0 ? 1 : 0);
    }
    
    return predictions;
}

int Perceptron::update_weights(const std::vector<std::vector<double>>& X, 
                               const std::vector<int>& y, 
                               const std::vector<int>& y_pred) {
    int misclassified = 0;
    
    // Update for each sample
    for (size_t i = 0; i < X.size(); ++i) {
        // Calculate error
        int error = y[i] - y_pred[i];
        
        // Update only if there's an error
        if (error != 0) {
            misclassified++;
            
            // Update bias
            bias_ += learning_rate_ * error;
            
            // Update weights
            for (size_t j = 0; j < weights_.size(); ++j) {
                weights_[j] += learning_rate_ * error * X[i][j];
            }
        }
    }
    
    return misclassified;
}

const std::vector<int>& Perceptron::get_errors() const {
    return errors_;
}

const std::vector<double>& Perceptron::get_weights() const {
    return weights_;
}

double Perceptron::get_bias() const {
    return bias_;
}

File: python/perceptron/perceptron.py
import numpy as np


class Perceptron:
    """
    A simple Perceptron implementation.
    
    The Perceptron is a linear binary classifier that learns weights 
    to separate data points into two classes.
    """
    
    def __init__(self, n_inputs, learning_rate=0.01, max_epochs=1000, random_state=None):
        """
        Initialize the Perceptron.
        
        Args:
            n_inputs (int): Number of input features
            learning_rate (float): Learning rate (eta)
            max_epochs (int): Maximum number of training epochs
            random_state (int): Seed for random number generator
        """
        self.learning_rate = learning_rate
        self.max_epochs = max_epochs
        
        # Set random seed for reproducibility
        if random_state:
            np.random.seed(random_state)
            
        # Initialize weights and bias
        self.weights = np.random.randn(n_inputs) * 0.01
        self.bias = 0.0
        
        # Training history
        self.errors = []
        
    def predict(self, X):
        """
        Make a prediction using the current model.
        
        Args:
            X (numpy.ndarray): Input features, shape (n_samples, n_inputs)
            
        Returns:
            numpy.ndarray: Predicted class labels (0 or 1)
        """
        # Calculate net input: w·x + b
        net_input = np.dot(X, self.weights) + self.bias
        
        # Apply step activation function
        return np.where(net_input >= 0, 1, 0)
    
    def _update_weights(self, X, y, y_pred):
        """
        Update weights based on prediction errors.
        
        Args:
            X (numpy.ndarray): Input features, shape (n_samples, n_inputs)
            y (numpy.ndarray): Target values, shape (n_samples,)
            y_pred (numpy.ndarray): Predicted values, shape (n_samples,)
        """
        # Calculate error
        errors = y - y_pred
        
        # Update weights using the perceptron learning rule
        # Δw = η * (y - ŷ) * x
        delta_w = self.learning_rate * np.dot(errors, X)
        delta_b = self.learning_rate * np.sum(errors)
        
        self.weights += delta_w
        self.bias += delta_b
        
        return np.sum(errors != 0)  # Return number of misclassifications
    
    def fit(self, X, y):
        """
        Train the perceptron on the given data.
        
        Args:
            X (numpy.ndarray): Training features, shape (n_samples, n_inputs)
            y (numpy.ndarray): Target values (0 or 1), shape (n_samples,)
            
        Returns:
            self: The trained model
        """
        self.errors = []
        
        # Training loop
        for epoch in range(self.max_epochs):
            # Make predictions
            y_pred = self.predict(X)
            
            # Update weights and count errors
            misclassified = self._update_weights(X, y, y_pred)
            self.errors.append(misclassified)
            
            # Stop if all samples are correctly classified
            if misclassified == 0:
                break
                
        return self